### ISSUE TITLE
OpenSSH is not doing the right thing with locale forwarding

### ISSUE THREAD

**Issue by jclulow** [TritonDataCenter/illumos-extra#141]
We have a patch for OpenSSH that is supposed to make it accept the **LANG** and **LC_\*** variables from the remote system by default:

https://github.com/TritonDataCenter/illumos-extra/blob/master/openssh/Patches/0028-Accept-and-send-LANG-and-LC_-environment-variables-b.patch

This is so that the current locale from your client system is correctly forwarded to the server system.  As far as I can tell, this used to work.  I just updated a system from **20240404T141941Z** to **20241128T004404Z**, though, and it is ostensibly not working anymore.

For example, SSH into a zone:

```
client $ LANG=en_AU.UTF-8 ssh -v server
OpenSSH_8.9p1 Ubuntu-3ubuntu0.10, OpenSSL 3.0.2 15 Mar 2022
debug1: Reading configuration data /home/jclulow/.ssh/config
...
debug1: Connection established.
...
debug1: Local version string SSH-2.0-OpenSSH_8.9p1 Ubuntu-3ubuntu0.10
debug1: Remote protocol version 2.0, remote software version OpenSSH_9.9
...
debug1: channel 0: new [client-session]
...
debug1: Sending environment.
debug1: channel 0: setting env LANG = "en_AU.UTF-8"

server $ locale
LANG=C
LC_CTYPE="C"
LC_NUMERIC="C"
LC_TIME="C"
LC_COLLATE="C"
LC_MONETARY="C"
LC_MESSAGES="C"
LC_ALL=
```

**Comment by jclulow** [comment]
I have done a few more experiments:

- **20240808T000421Z** with **OpenSSH_9.8p1** appears to have the problem
- **20240701T205528Z** with **OpenSSH_9.7p1** _works correctly!_

To reproduce without reconfiguring your local workstation to have `SendEnv LC_* LANG`, and without needing to use a SmartOS client, you can:

```
$ ssh -o SetEnv=LANG=en_AU.UTF-8 sigma locale
LANG=en_AU.UTF-8
LC_CTYPE="en_AU.UTF-8"
LC_NUMERIC="en_AU.UTF-8"
LC_TIME="en_AU.UTF-8"
LC_COLLATE="en_AU.UTF-8"
LC_MONETARY="en_AU.UTF-8"
LC_MESSAGES="en_AU.UTF-8"
LC_ALL=
```

Note that depending on your zone image, or your global zone configuration, the locale might naturally either be **C** or **en_US.UTF-8**, so it's important to use a non-default locale like **en_AU.UTF-8** to ensure it's passing through correctly.

**Comment by danmcd** [comment]
It's getting sent over the protocol, but the server is just not having it:

```
debug1: Sending environment.
debug1: channel 0: setting env LANG = "en_AU.UTF-8"
debug1: pledge: fork

  ,---.                   |     ,---. ,---.
  `---. ,-.-. ,---. ,---. |---  |   | `---.  base-64-lts
      | | | | ,---| |     |     |   |     |  21.4.1
  `---' ` ' ' `---' `     `---' `---' `---'

smartos-build-3(~)[0]% echo $LANG
C
smartos-build-3(~)[0]% logout
```

**Comment by danmcd** [comment]
OpenSSH 9.8p1 introduce the split process from sshd of `sshd-session`.

```
smartos-build-3(~)[0]% ptree $$
128164 zsched
  128624 /sbin/init
    129419 /usr/lib/ssh/sshd
      129460 /usr/lib/ssh/sshd-session -R
        129465 /usr/lib/ssh/sshd-session -R
          129466 -tcsh
            886036 ptree 129466
smartos-build-3(~)[0]% 
```

But I have a `setenv LANG C` in my .tcshrc.  Lemme remove it...

```
kebe(~)[0]% ( setenv LANG en_AU.UTF-8 ; ssh build3 "ssh -V ; grep PermitUserEnv /etc/ssh/sshd_config ; uname -a ; printenv LANG" )
(danmcd@build3) Password: 
OpenSSH_9.9p1, OpenSSL 3.0.15 3 Sep 2024
PermitUserEnvironment yes
SunOS smartos-build-3 5.11 joyent_20241202T163610Z i86pc i386 i86pc
en_AU.UTF-8
kebe(~)[0]% 
```

Hmmm.

**Comment by jclulow** [comment]
So it's definitely not the client, because I tried the same client against like five different SmartOS versions lined up with the OpenSSH upgrades.  I think it must be an artefact of the new split process stuff in 9.8p1.

I will note, though, that _another_ (not SSH-specific?) difference I found when poking at SmartOS and, say, Helios/OmniOS is that LANG appears to be set in the environment of more processes on SmartOS.

SmartOS, where apparently `LANG=C` is set in the environment of `sshd`:

```
 $ ptree $$ | awk '{ print $1 }' | xargs pfexec pargs -e  | egrep '^[0-9]|LANG'
4441: zsched
5707:   /sbin/init
8454:   /usr/lib/ssh/sshd
envp[1]: LANG=C
9604:   /usr/lib/ssh/sshd -R
envp[1]: LANG=C
9606:   /usr/lib/ssh/sshd -R
envp[1]: LANG=C
9607:   -bash
pargs: cannot examine 12481: no such process or core file
envp[13]: LANG=en_US.UTF-8
pargs: cannot examine 12482: no such process or core file
12483:  xargs pfexec pargs -e
envp[13]: LANG=en_US.UTF-8
12484:  egrep ^[0-9]|LANG
envp[13]: LANG=en_US.UTF-8
```

And Helios, where it does not:

```
 $ ptree $$ | awk '{ print $1 }' | xargs pfexec pargs -e  | egrep '^[0-9]|LANG'
517:    /usr/sbin/sshd
10060:  /usr/sbin/sshd -R
10062:  /usr/sbin/sshd -R
10063:  -bash
pargs: cannot examine 10076: no such process or core file
pargs: cannot examine 10077: no such process or core file
envp[12]: LANG=en_US.UTF-8
10078:  xargs pfexec pargs -e
envp[12]: LANG=en_US.UTF-8
```

Perhaps 9.8p1 is screwing up imported the environment from the client _if it's already set in **sshd**?_  I'm not sure where the `LANG=C` comes from, there, but I suspect it might be some interaction between `/etc/default/init`, which in this zone image appears to contain `LANG=C`, and SMF.

Regardless, sshd used to correctly override this even when already set in 9.7p1, and apparently no longer does in our 9.8p1?

**Comment by jclulow** [comment]
FWIW, **PermitUserEnvironment** is not required: that's about whether `~/.ssh/environment`, or `environment=` options in the authorised keys file, are processed.

I believe the option that allows environment variables to be passed through the session from the client is **AcceptEnv**, which on illumos systems defaults to accepting anything matching **LANG** or **LC_\***, to replace the original SunSSH "locale negotiation" facility.

**Comment by danmcd** [comment]
Got it.

**Comment by danmcd** [comment]
https://smartos.org/bugview/OS-8607

**Comment by danmcd** [comment]
PR is #142

**Comment by danmcd** [comment]
Fixed in #142
